File: sdk/ts/src/client.test.ts
/**
 * Self-pair integration test: builds and runs the real cmd/relayly server binary
 * (not an in-process test double) and drives two sdk/ts RelaylyClients through it
 * end to end — register, connect, pair, a real Noise XX handshake, and bidirectional
 * encrypted delivery. This is the "each SDK against itself" leg of the interop matrix
 * (docs/tasks/02-sdks-and-interop.md), landed early since it directly de-risks this
 * PR the same way sdk/go's client_test.go does.
 *
 * Node has no global WebSocket on two of the three Node versions this repo's CI
 * matrix tests (18, 20; only 22 has it natively), so this file polyfills
 * globalThis.WebSocket with the `ws` package before importing RelaylyClient — a
 * test-only devDependency, not a runtime dependency of the SDK.
 */
import { WebSocket as NodeWebSocket } from 'ws';

// eslint-disable-next-line @typescript-eslint/no-explicit-any
(globalThis as any).WebSocket = NodeWebSocket;

import { describe, expect, it } from 'vitest';
import { execFile, spawn, type ChildProcess } from 'node:child_process';
import { promisify } from 'node:util';
import { mkdtemp, rm } from 'node:fs/promises';
import { tmpdir } from 'node:os';
import { dirname, join } from 'node:path';
import { fileURLToPath } from 'node:url';
import { createServer } from 'node:net';

const execFileAsync = promisify(execFile);
const currentDir = dirname(fileURLToPath(import.meta.url));

async function repoRoot(): Promise<string> {
  // sdk/ts/src is three levels under the repo root.
  return join(currentDir, '..', '..', '..');
}

async function buildRelayServer(): Promise<string> {
  const root = await repoRoot();
  const dir = await mkdtemp(join(tmpdir(), 'relayly-server-'));
  const binPath = join(dir, 'relayly-server');
  await execFileAsync('go', ['build', '-o', binPath, './cmd/relayly'], { cwd: root });
  return binPath;
}

async function freePort(): Promise<number> {
  return new Promise((resolve, reject) => {
    const srv = createServer();
    srv.listen(0, '127.0.0.1', () => {
      const address = srv.address();
      const port = typeof address === 'object' && address ? address.port : 0;
      srv.close(() => resolve(port));
    });
    srv.on('error', reject);
  });
}

interface RunningServer {
  baseUrl: string;
  wsUrl: string;
  stop: () => Promise<void>;
}

async function startRelayServer(binPath: string): Promise<RunningServer> {
  const port = await freePort();
  const dbDir = await mkdtemp(join(tmpdir(), 'relayly-db-'));
  const dbPath = join(dbDir, 'relayly.db');

  const child: ChildProcess = spawn(binPath, ['start', '--host', '127.0.0.1', '--port', String(port), '--db.path', dbPath], {
    stdio: ['ignore', 'ignore', 'inherit'],
  });

  const baseUrl = `http://127.0.0.1:${port}`;
  const deadline = Date.now() + 10_000;
  while (Date.now() < deadline) {
    try {
      const resp = await fetch(`${baseUrl}/health`);
      if (resp.ok) break;
    } catch {
      // not up yet
    }
    await new Promise((r) => setTimeout(r, 50));
  }

  return {
    baseUrl,
    wsUrl: `ws://127.0.0.1:${port}/ws`,
    stop: async () => {
      child.kill();
      await rm(dbDir, { recursive: true, force: true });
    },
  };
}

interface DeviceCreds {
  device_id: string;
  device_token: string;
}

async function registerDevice(baseUrl: string, name: string): Promise<DeviceCreds> {
  const resp = await fetch(`${baseUrl}/api/v1/devices`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ name }),
  });
  if (!resp.ok) throw new Error(`registering device ${name}: status ${resp.status}`);
  return (await resp.json()) as DeviceCreds;
}

describe('RelaylyClient self-pair integration', () => {
  it('registers, connects, pairs, and exchanges encrypted messages both ways', async () => {
    const { RelaylyClient } = await import('./client.js');
    const { generateKey } = await import('./crypto.js');
    const { InMemoryPeerKeyStore } = await import('./peerStore.js');

    const binPath = await buildRelayServer();
    const server = await startRelayServer(binPath);
    try {
      const devA = await registerDevice(server.baseUrl, 'device-a');
      const devB = await registerDevice(server.baseUrl, 'device-b');

      const clientA = new RelaylyClient(server.wsUrl, {
        deviceId: devA.device_id,
        deviceToken: devA.device_token,
        keyPair: generateKey(),
        peerStore: new InMemoryPeerKeyStore(),
      });
      const clientB = new RelaylyClient(server.wsUrl, {
        deviceId: devB.device_id,
        deviceToken: devB.device_token,
        keyPair: generateKey(),
        peerStore: new InMemoryPeerKeyStore(),
      });

      await clientA.connect();
      await clientB.connect();

      const code = await clientA.requestPairCode();
      expect(code.shortCode).toHaveLength(6);

      const waitForA = clientA.waitForPairing();
      const peerB = await clientB.acceptPair(code.shortCode);
      const peerA = await waitForA;

      expect(peerA.id).toBe(devB.device_id);
      expect(peerB.id).toBe(devA.device_id);

      const messageFromB = new Promise<string>((resolve) => {
        clientB.on('message', (msg) => resolve(msg.payload));
      });
      await clientA.send(peerA.id, 'hello from A');
      expect(await messageFromB).toBe('hello from A');

      const messageFromA = new Promise<string>((resolve) => {
        clientA.on('message', (msg) => resolve(msg.payload));
      });
      await clientB.send(peerB.id, 'hello from B');
      expect(await messageFromA).toBe('hello from B');

      clientA.disconnect();
      clientB.disconnect();
    } finally {
      await server.stop();
    }
  }, 30_000);
});

File: sdk/go/client_test.go
package relayly_test

// Self-pair integration test: builds and runs the real cmd/relayly server binary
// (not an in-process test double) and drives two sdk/go Clients through it end to
// end — register, connect, pair, a real Noise XX handshake, and bidirectional
// encrypted delivery. This is the "each SDK against itself" leg of the interop matrix
// (docs/tasks/02-sdks-and-interop.md), landed early here since it directly de-risks
// this PR: PR 2's own migration bug was only ever caught by running the real binary,
// never by unit tests against a fake.

import (
	"context"
	"encoding/json"
	"fmt"
	"net"
	"net/http"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
	"testing"
	"time"

	relayly "github.com/NIKX-Tech/relayly/sdk/go"
)

func buildRelayServer(t *testing.T) string {
	t.Helper()
	repoRoot, err := filepath.Abs("../..")
	if err != nil {
		t.Fatalf("resolving repo root: %v", err)
	}
	if _, err := os.Stat(filepath.Join(repoRoot, "cmd", "relayly")); err != nil {
		t.Fatalf("cmd/relayly not found at %s (expected sdk/go to be two levels under the repo root): %v", repoRoot, err)
	}

	binPath := filepath.Join(t.TempDir(), "relayly-server")
	cmd := exec.Command("go", "build", "-o", binPath, "./cmd/relayly")
	cmd.Dir = repoRoot
	if out, err := cmd.CombinedOutput(); err != nil {
		t.Fatalf("building relay server: %v\n%s", err, out)
	}
	return binPath
}

func freePort(t *testing.T) int {
	t.Helper()
	l, err := net.Listen("tcp", "127.0.0.1:0")
	if err != nil {
		t.Fatalf("finding a free port: %v", err)
	}
	defer l.Close()
	return l.Addr().(*net.TCPAddr).Port
}

func startRelayServer(t *testing.T, binPath string) string {
	t.Helper()

	port := freePort(t)
	dbPath := filepath.Join(t.TempDir(), "relayly.db")

	cmd := exec.Command(binPath, "start",
		"--host", "127.0.0.1",
		"--port", fmt.Sprintf("%d", port),
		"--db.path", dbPath,
	)
	cmd.Stdout = os.Stderr
	cmd.Stderr = os.Stderr
	if err := cmd.Start(); err != nil {
		t.Fatalf("starting relay server: %v", err)
	}
	t.Cleanup(func() {
		_ = cmd.Process.Kill()
		_ = cmd.Wait()
	})

	baseURL := fmt.Sprintf("http://127.0.0.1:%d", port)
	deadline := time.Now().Add(10 * time.Second)
	for time.Now().Before(deadline) {
		resp, err := http.Get(baseURL + "/health")
		if err == nil {
			resp.Body.Close()
			if resp.StatusCode == http.StatusOK {
				return baseURL
			}
		}
		time.Sleep(50 * time.Millisecond)
	}
	t.Fatal("relay server did not become healthy in time")
	return ""
}

type createDeviceResponse struct {
	DeviceID    string `json:"device_id"`
	DeviceToken string `json:"device_token"`
}

func registerDevice(t *testing.T, baseURL, name string) createDeviceResponse {
	t.Helper()
	body := strings.NewReader(fmt.Sprintf(`{"name":%q}`, name))
	resp, err := http.Post(baseURL+"/api/v1/devices", "application/json", body)
	if err != nil {
		t.Fatalf("registering device %s: %v", name, err)
	}
	defer resp.Body.Close()
	if resp.StatusCode != http.StatusOK {
		t.Fatalf("registering device %s: status %d", name, resp.StatusCode)
	}
	var out createDeviceResponse
	if err := json.NewDecoder(resp.Body).Decode(&out); err != nil {
		t.Fatalf("decoding device response: %v", err)
	}
	return out
}

func TestSelfPair_PairAndExchange(t *testing.T) {
	if testing.Short() {
		t.Skip("skipping self-pair integration test in -short mode")
	}

	binPath := buildRelayServer(t)
	baseURL := startRelayServer(t, binPath)
	wsURL := strings.Replace(baseURL, "http://", "ws://", 1) + "/ws"

	devA := registerDevice(t, baseURL, "device-a")
	devB := registerDevice(t, baseURL, "device-b")

	keyA, err := relayly.GenerateKey()
	if err != nil {
		t.Fatalf("GenerateKey A: %v", err)
	}
	keyB, err := relayly.GenerateKey()
	if err != nil {
		t.Fatalf("GenerateKey B: %v", err)
	}

	ctx, cancel := context.WithTimeout(context.Background(), 20*time.Second)
	defer cancel()

	clientA, err := relayly.Connect(ctx, wsURL, relayly.Options{
		DeviceID:      devA.DeviceID,
		DeviceToken:   devA.DeviceToken,
		PrivateKey:    keyA,
		PeerStorePath: filepath.Join(t.TempDir(), "a-peers.json"),
	})
	if err != nil {
		t.Fatalf("connect A: %v", err)
	}
	defer clientA.Close()

	clientB, err := relayly.Connect(ctx, wsURL, relayly.Options{
		DeviceID:      devB.DeviceID,
		DeviceToken:   devB.DeviceToken,
		PrivateKey:    keyB,
		PeerStorePath: filepath.Join(t.TempDir(), "b-peers.json"),
	})
	if err != nil {
		t.Fatalf("connect B: %v", err)
	}
	defer clientB.Close()

	code, err := clientA.RequestPairCode(ctx)
	if err != nil {
		t.Fatalf("RequestPairCode: %v", err)
	}

	var peerA *relayly.Peer
	var waitErr error
	pairDone := make(chan struct{})
	go func() {
		defer close(pairDone)
		peerA, waitErr = code.Wait(ctx)
	}()

	// AcceptPair blocks until the Noise handshake (and §7 pin/cross-check) actually
	// completes, so peerB is immediately usable with Send.
	peerB, err := clientB.AcceptPair(ctx, code.Short)
	if err != nil {
		t.Fatalf("AcceptPair: %v", err)
	}
	<-pairDone
	if waitErr != nil {
		t.Fatalf("A's code.Wait: %v", waitErr)
	}
	if peerA == nil || peerA.ID != devB.DeviceID {
		t.Fatalf("unexpected peerA: %+v", peerA)
	}
	if peerB.ID != devA.DeviceID {
		t.Fatalf("unexpected peerB: %+v", peerB)
	}

	if err := clientA.Send(ctx, peerA.ID, []byte("hello from A")); err != nil {
		t.Fatalf("A send: %v", err)
	}
	select {
	case msg := <-clientB.Messages():
		if string(msg.Payload) != "hello from A" {
			t.Errorf("B got %q", msg.Payload)
		}
		if msg.From != devA.DeviceID {
			t.Errorf("B's message.From: want %s, got %s", devA.DeviceID, msg.From)
		}
	case <-time.After(5 * time.Second):
		t.Fatal("B did not receive A's message")
	}

	if err := clientB.Send(ctx, peerB.ID, []byte("hello from B")); err != nil {
		t.Fatalf("B send: %v", err)
	}
	select {
	case msg := <-clientA.Messages():
		if string(msg.Payload) != "hello from B" {
			t.Errorf("A got %q", msg.Payload)
		}
	case <-time.After(5 * time.Second):
		t.Fatal("A did not receive B's message")
	}
}
